### DISCUSSION TITLE
Choose dictionary entry when learning new kanji

### DISCUSSION THREAD

**Discussion by hermi362** [fabd/kanji-koohii#239] [1. Feedback & Feature Requests]
Firstly, thank you so much for making this site, I've been using it on a daily basis since 2017.

When I'm in the study page learning a new kanji, I very often would like to add a dictionary entry to a kanji I am learning. Having an example word is helpful to me.
As far as I can work out, this is only possible during a review session, not while studying/browsing.
I wait until the kanji appears in the review session and then open the dictionary and assign an example word.

It would be nice if there was a way to assign an example word from the study page (I suppose it would mean that kanji gets automatically added to list of learned kanji)

![image](https://user-images.githubusercontent.com/5420254/149685893-d482ae4d-c50a-48fc-a447-49c0ea97e4a4.png)

**Comment by fabd** [comment]
> It would be nice if there was a way to assign an example word from the study page

I agree. It was kind of a logical step that I always intended to do.

It's just kind of stuck in brainstorming stage, with my attention going to other things. So thanks for reminding me about it!

It's not straightforward to implement because I can not cache this part of the page, if I show the user's selected example word. But it is solvable.

What I had in mind is to display the user's selected word (whether it was selected here, or during review), below the story, even if the Dictionary is not currently opened/expanded. It's just more database queries and thinking about how to optimize this.

> (I suppose it would mean that kanji gets automatically added to list of learned kanji)

So that's the interesting question, which is partly why I was undecided. Dictionary selected items are called " vocab picks" in the code. There is a separate table for them, so they do not require a flashcard. They are simply a link between a unique kanji, and an entry from JMDICT. So à priori in my initial implementation "plan", it would not necessitate a flashcard -- and when the user adds a flashcard afterwards, the selected vocab remains in place.

For a lot of people that don't use the SRS, it lets them basically "highlight" one dictionary word per Study page. But it doesn't do anything besides when you revisit the page.

Besides the SRS I figured I could perhaps have a separate page where you can get a list of your vocab, similar to the Flashcard List page. Could be useful for studying.

I haven't thought of anything else for now. 

So yeah to summarize it's definitely something I intended to do, but it's still kind of unclear for me how I make it fit into the site's purpose -- as I feel adding a vocab SRS is outside the scope of the site -- so I need something to make it "click" and I don't know what it is.

Maybe just the ability for users to retrieve the list of their selected vocab is it,. I'm not sure.

**Comment by hermi362** [comment]
Thank you for the comments, Fabrice.

It is really quite a minor matter with very small impact for me, so please don't let my post influence the priority of adding this feature.

It is just a matter of habit -- the vocab picks are available when reviewing new kanjis, which is convenient enough.

Overall I think the site is a great help and well thought-out. Thank you for creating it!

**Comment by Dragonforge** [comment]
Hi!
Very interesting question indeed.
I've another question related to "vocab picks".

_"Dictionary selected items are called " vocab picks" in the code. There is a separate table for them, so they do not require a flashcard. They are simply a link between a unique kanji, and an entry from JMDICT. So à priori in my initial implementation "plan", it would not necessitate a flashcard -- and when the user adds a flashcard afterwards, the selected vocab remains in place."_

I'm using another SRS app to learn vocabulary, and i would really be interested in exporting the vocabulary i've picked for kanjis.
So would it be possible to export the "vocab picks" ?
That way, i would easily export the vocabulary to the SRS app and work on these words.

Best,
Jacques

**Reply by fabd** [reply]
Hmm yes I am sorry there is no export of that particular table atm, as far as I remember.

If you or another dev is particularly motivated, you could add an "Export Vocab" option in the [Manage > Export](https://kanji.koohii.com/manage/export) section, at the bottom of the sidebar there. Adding an action in symfony is pretty straightforward, by copying the "Export flashcards" code.

On further thought - I believe atm there is a hardcoded limit of 1 vocab per card - so this could potnetially just be added to the flashcards export in the SQL: query, one mopre join and an additional column in the CSV.

I'm sorry I'm kinda swamped atm, need to replace my mobo "CMOS", need  to setup ethernet because my PC wifi is so laggy , buy a switch, cables. Oh yeah, and now my docker also gives me permission issues so php can not write to files - which happened out of nowhere and I´;ll have to upgrade the Dockerfile and find a solution. So frustrating :cry:  ..  I swear this never ends.

Been wanting to sit down and code for weeks now >_<   

If someone puts up a PR I should be able to merge it.

**Reply by Dragonforge** [reply]
Hi! Mark has sent a PR :-)
I hope your problems will find a solution too. Keep it up!

**Reply by fabd** [reply]
Hi @Dragonforge  I saw it yesterday, was able to sit down today and the site is updated with Mark's PR. Cheers!

**Reply by Dragonforge** [reply]
Thanks a lot!

**Comment by Dragonforge** [comment]
@fabd Hi ! Thanks for mentioning us in your post ^^
Concerning vocabulary, what's really missing for me is a way to add vocabulary words for kanjis that do not possess any "dictionary word"
This forces me to open other apps, so it becomes a necessity to be "out of" kanji koohii.
What i would really like, is a way to add one word for kanjis that don't have any words associated (and which are not radical-only, or fragmented kanji-part). Can it be done ?

**Reply by fabd** [reply]
> Can it be done ?

Probably not. The code uses the basic assumption which is the link between a unique flashcard and a dictionary entry. Unique flashcard id is the unicode code point - while the dictionary entry is JMDICT's `ent_seq` value which is a unique entry id.

I also made the choice to "only" include the common words - which is 16k + words as explained here: https://kanji.koohii.com/learnmore#dictionary-sources . In order to allow used to add ANY JMDICT existing entry - I´d have to parse and store 170k+ words in the database.

The more flexible and easier option would be something like allowing to have custom notes on each card. I thought about it, but fitting everything in a smartphone screen is always a PITA and not very motivating for me.

**Reply by Dragonforge** [reply]
Hi! 
I do understand your choices although i am wondering...

The interface is great as it is! So don't worry about it or note-taking (we already have stories available).

If you parse and store 170k+ in the database, would it change something in the code ? 

`This is done in order to provide meaningful results for Koohii users. The dictionary would otherwise show rare, archaic and obsolete words as well as place names from a total of 170,000+ entries which you'll probably never use even after being fluent in Japanese.`

I quite agree, but could there be an option : show only 17000 first entries or show "whole" dictionary... Would there be a bandwith issue ?

I ask, not because i want to be fluent, but because i find associating a kanji with a keyword and a key-vocabulary quite powerful.
Take for example the kanji 孔 (in RTK)
I have a hard time remembering it on its own,
But if i see the key-vocabulary : 鼻孔 (nostril), it really "sticks". But 鼻孔 does not fit in the 16k+ words....

So if it can't be done, it can't be... But at least, i'll have asked :-)

**Reply by ssjoel3k** [reply]
I've also been wondering this as well.
I realize parsing and storing 170k+ words in the database might be too much, but it would be nice to have at least one entry for each kanji.
It'd be great if we took the top 1 or 2 most frequent words for each kanji and added them to the database.
Given that there are 3000 kanji between RTK 1&3, that would only be an extra 3000-6000 records.
Name-specific kanji might not need entries I suppose (although an example name might be nice).

The other alternative would be an option to add your own vocab which would be nice as well.
As @Dragonforge mentioned, kanji like 孔 are very hard to remember by themselves, so being able to add vocab like 鼻孔 would help a lot.
